exercise-app.js

```diff
@@ -28,13 +19,59 @@ const Weather = ({ temp, description, icon }) => (
 );
 
 
-const App = () => (
+class App extends React.Component {
+
+  constructor(props) {                // where do we hook up these props to ? where does this prop data come from?
+  super(props);                       // what is the purpose of this? why do we need to call super?
+  this.state = {
+  temp: 0,
+  description: 'not loaded yet',
+  icon: `http://openweathermap.org/img/w/10d.png`};  // always intialize a state? is the state to be updated by passing in props?
+  // this.reshapeData = this.reshapeData.bind(this)  // when do we need to bind and when not (I've seen both...)
+  this.render()                       // was i supposed call render here? (I didn't and the card didn't show for a bit upon load)
+}
+
+ getWeather() {
+    return fetch(`http://api.openweathermap.org/data/2.5/weather?id=6167865&appid=581522ec0b69c0bcc097344d84601245&units=metric`)
+      .then(res => {
+        // console.log(37,res && res.json())     // when I uncomment this... I get error "promise already used" why?
+        return res.json()})
+      .then((dt)=>{
+        console.log(40, dt)
+        this.reshapeData(dt)
+      })
+  }
+
+   reshapeData(data) {
+    console.log(data)
+    this.setState({
+      temp: `${data.main.temp} timestamp: ${+ new Date()}`,       // I added random time stamp for easier debugging... plz ignore
+      description: data.weather[0].description,
+      icon: `http://openweathermap.org/img/w/${"09d"}.png` 
+    }
+    )
+  }
+
+componentDidMount() {
+    this.timerID = setInterval(() => this.getWeather(), 15000); // is this.timerID a prop or a state?
+
+  }
+
+  componentWillUnmount() {
+  clearInterval(this.timerID);                 // how does it know when to unmount? what is I didn't clear the interval?                                   // typically do I need to try and clean up in componentWillUnmount
```

Comment:
When you remove the component from the DOM. So, for example if you navigate to a new route or you had some view logic which controlled the rendering of the card: `{ showCard ? <WeatherCard .../> : null }`, etc.

> what if I didn't clear the interval? 
This will cause memory leaks. Imagine you had a button to show or hide the weather card. Every time the user clicks show the component mounts and you start a timer and start fetching data. Then the user clicks hide and the time is still alive. Then the user clicks show again and now another timer was created. We now have 2 timers. Now imagine if the user click show/hide 100s of times. Each time we create a timer which takes up memory (RAM).